File: crypto/demo/gcm.js
const crypto = require('crypto')

/**
 * {}
*/
function encrypt(baseId, plaintext, key) {
    let nonce = crypto.randomBytes(12)

    const cipher = crypto.createCipheriv('aes-256-gcm', key, nonce)

    // gcm sets the value used for the additional authenticated data (AAD) input parameter.
    cipher.setAAD(baseId)

    let encrypted = Buffer.concat([cipher.update(plaintext, 'utf8'), cipher.final()])
    let tag = cipher.getAuthTag()
    let ciphertext = Buffer.concat([nonce, encrypted, tag]).toString('hex')
    console.log(ciphertext)
    return ciphertext
}

function decrypt(baseId, encryptedText, key) {
    let bData = Buffer.from(encryptedText, 'hex')

    let nonce = bData.slice(0, 12)
    let tag = bData.slice(bData.length - 16, bData.length)
    let text = bData.slice(12, bData.length - 16)

    const decipher = crypto.createDecipheriv('aes-256-gcm', key, nonce)
    decipher.setAuthTag(tag)
    decipher.setAAD(baseId)
    try {
        let receivedPlaintext = decipher.update(text, 'binary', 'utf8') + decipher.final('utf8')
        console.log(receivedPlaintext)
        return receivedPlaintext
    } catch (error) {
        console.error('Authentication failed!')
        return
    }
}

module.exports = {
    encrypt,
    decrypt,
}

File: crypto/demo/gcm.go
package main

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/hex"
	"io"
)

func encrypt(baseId string, data []byte, key []byte) (string, error) {
	plainText := []byte(data)

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}

	nonce := make([]byte, 12)
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return "", err
	}

	aesGCM, err := cipher.NewGCM(block)
	if err != nil {
		panic(err.Error())
	}

	cipherText := aesGCM.Seal(nil, nonce, plainText, []byte(baseId))

	return hex.EncodeToString(nonce) + hex.EncodeToString(cipherText), nil
}

func decrypt(baseId string, msg string, key []byte) (io.Reader, error) {
	data, err := hex.DecodeString(msg)
	if err != nil {
		return nil, err
	}

	nonce := data[:12]
	cipherText := data[12:]

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}

	aesGCM, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	plainText, err := aesGCM.Open(nil, nonce, cipherText, []byte(baseId))
	if err != nil {
		return nil, err
	}

	return bytes.NewBuffer(plainText), nil
}
